### PR TITLE
GPS Fix estimation (dead reconing, RTH without GPS fix) for fixed wing

### PR DESCRIPTION
**Pull request by RomanLut**
# GPS Fix estimation (dead reconing, RTH without GPS fix) for fixed wing

Video demonstration

[IMAGE ALT TEXT HERE](https://www.youtube.com/watch?v=wzvgRpXCS4U)

There is possibility to allow plane to estimate it's position when GPS fix is lost.
The main purpose is RTH without GPS.
It works on fixed wing only.

Up-to-date description:
https://github.com/RomanLut/inav/blob/submit-gps-fix-estimation/docs/GPS_fix_estimation.md

Requires configurator change:
https://github.com/iNavFlight/inav-configurator/pull/1669

### CODE REVIEW

**Comment by RomanLut** [comment]
Tested in simulator only :(

**Comment by RomanLut** [comment]
PR is prepared to be merged after https://github.com/iNavFlight/inav/pull/8268

**Comment by JulioCesarMatias** [comment]
I think some problems occurred during your push... Because you have several files modified, with the HIL mode included... You have 2 options:

1 - Close this PR and open a new one with only the modifications of this new resource.

2 - Sync the branch of this PR with the Master, and delete the files from the HIL mode.

As there were many files changed, I myself couldn't read the algorithm and understand what you changed to add the feature (maybe it's even laziness on my part)

**Comment by JulioCesarMatias** [comment]
yeah... this PR is hard to analyze now, I'll leave it for the future

**Comment by RomanLut** [comment]
Problems will vanish once  https://github.com/iNavFlight/inav/pull/8268 and https://github.com/iNavFlight/inav/pull/8346 are merged into master.

For now you can use https://github.com/RomanLut/inav/pull/10/files for analisys.

**Comment by JulioCesarMatias** [comment]
the "dead-reckoning" part of the coordinates and velocity also exists in a new AHRS that I'm working on bringing to iNav... Glad we had the same idea ^^

**Comment by JulioCesarMatias** [comment]
the "BOXGPSOFF" option should only be kept during testing before that one day it is approved to merge... It will be good help for Dev's testing... When this feature has a final version, I recommend removing the option to force the GPS failure (I don't think you want users bothering you and saying that your algorithm crashes their planes)

**Comment by DzikuVx** [comment]
@RomanLut please update docs and resolve conflicts please

**Comment by RomanLut** [comment]
> @RomanLut please update docs and resolve conflicts please

done.

**Comment by breadoven** [comment]
Does the wind estimator only seem to work if you fly in circles ? This is typically how wind estimators work with software for paragliders/hang gliders etc, either rely on flying in circles (thermaling) or flying in zig zags was another option.

**Comment by RomanLut** [comment]
Zigzag should work too.

Wind estimator does estimation when attititude changes by ~10 degrees. Then estimated wind vector is updated with 5% of computed value. ~36 estimations in circle is enough to get good estimation.

https://drive.google.com/file/d/0ByvTkVQo3tqXVzBYQUZicUNvbEE/view?resourcekey=0-s-7v3PgOUrZt7fI6JNlu-Q

**Comment by shota3527** [comment]
@RomanLut 
> Cruise airspeed is specified in m/s.

To convert km/h to m/s, multiply by 27.77.

Example: 100 km/h = 100 * 27.77 = 2777 m/s

is this cm/s or m/s ?

**Comment by RomanLut** [comment]
> is this cm/s or m/s ?

You are right, it should be cm/s, updated documentation. Thanks.

**Comment by b14ckyy** [comment]
Hi @RomanLut 
I want to real flight test this feature in the next days. But it has conflicts. Can you solve them so I can put it in a test build?

**Comment by RomanLut** [comment]
1) Fixed conflicts with current master branch
2) Removed "GPS Off" box
3) Implemented "Disabled GPS Sensor Fix" logic condition https://github.com/iNavFlight/inav-configurator/pull/1669

**Comment by RomanLut** [comment]
> Hi @RomanLut I want to real flight test this feature in the next days. But it has conflicts. Can you solve them so I can put it in a test build?

Quick reminder:
1) set inav_allow_gps_fix_estimation=ON
2) set fw_reference_airspeed=2777  
Set for your wing. 2777 is valid for FPV SurfWing.
3) save
4) setup "Disable GPS Sensor fix" functionality in programming tab.
5) Get used with feature in simulator
6) Fly with stable RC and Video Link while testing!

More details in docs: 
https://github.com/RomanLut/inav/blob/submit-gps-fix-estimation/docs/GPS_fix_estimation.md

**Comment by RomanLut** [comment]
Yes code will use pitot tube's airspeed if pitot tube is available.

**Comment by b14ckyy** [comment]
@RomanLut during Simulator testing I found something weird. 
If I disable the GPS fix, the airspeed also changes to the fw_reference_airspeed and is not updated by the Airspeed sensor anymore. 

Right now I have set the airspeed sensor for the simulator to VIRTUAL as I thought that the HITL Plugin then simulates a physical Airspeed sensor. Is this wrong?

**Comment by b14ckyy** [comment]
Okay nevermind. Mea culpa. 
I forgot to turn on Pitot Simulation in the Plugin settings and it was always running from actual Virtual Pitot data.

**Comment by RomanLut** [comment]
> Okay nevermind. Mea culpa. I forgot to turn on Pitot Simulation in the Plugin settings and it was always running from actual Virtual Pitot data.

Yes, without GPS fix, Virtual pitot falls back to fw_reference_airspeed.

**Comment by b14ckyy** [comment]
So in the SIM it works really well. if Wind Speed is properly estimated, even with some gusts, I can run my full autoland mission on GPS estimation and still touch down within the runway boundaries. 

But I have a suggestion for a feature change. For set `inav_allow_gps_fix_estimation` I suggest to add a 3rd option: ON_FS
So as soon as the GPS fix is lost, no matter what, the plane should initiate a RTH if that setting is set. having this feature on while in a WP mission without previously properly circle to get the correct wind speed, continuing the mission could be dangerous. Going home on a direct path massively raises the chance to get the plane back.

Currently the User would not be made aware of a GPS loss when it happens if the FPV feed is not monitored.

**Comment by RomanLut** [comment]
> But I have a suggestion for a feature change. For set `inav_allow_gps_fix_estimation` I suggest to add a 3rd option: ON_FS So as soon as the GPS fix is lost, no matter what, the plane should initiate a RTH if that setting is set.

Well, I am not yet sure we should pack multiple subjects into single setting because inav_allow_gps_fix_estimation has nothing to do with missions. It is about enabling or disabling core feature.

"FS_ON" (or longer "ON_WITH_FAILSAFE_MISSION) would mean:
1) enable GFX Fiix esimation AND
2) on lost GPS fix event, trigger failsafe action (as configured by failsafe_procedure), if flight mode is WP, with hardcoded delay.

Maybe there  should be separate options for mission configuration, like failsafe_mission_on_gps_fix_estimation_procedure, failsafe_mission_on_gps_fix_estimation_delay.

On the other hand, I would stop production options but evolve Programming tab for such behavior scripts instead.
You already can detect GPS fix estimation (GPS_Sats =99), wait for delay, and trigger RTH or Failsafe (with RC_Channels_Override) in programming tab.

To make this more convient, I would adjust GPS_Fix operand with 0-no fix, 1-has fix, 2-estimation, and implement specific logic condition to trigger Failsave, RTH, Land, or Disarm actions without RC_Channels_Override trick.

I would like to hear core developers opinion on this because the one who accepts push requests does such decisions :)

**Comment by RomanLut** [comment]
What I am trying to say is that we should not add more options to inav_allow_gps_fix_estimation to control multiple things.
1) inav_allow_gps_fix_estimation should control gfx fix estimation only - on/off
2) we have an option to contol whether mission will switch to to failsafe on RC link lost: failsafe_mission_delay
So there should be more options like:
failsafe_mission_gps_fix_estimation_delay - controls whether massion swithes to failsafe mode on gps fix estimation 
failsafe_mission_low_battery_delay - controls whether massion swithes to failsafe on low battery etc.

**Comment by b14ckyy** [comment]
I agree on `failsafe_mission_gps_fix_estimation_delay`, this sounds like a good option. But maybe call it in general `gps_fix_estimation_failsafe_delay` and apply it to any navigation mode. So if GPS Fix Estimation has to take over due to a fix loss, the plane can failsafe and RTH in any GPS assisted mode unless the user takes back over with stick input or disables navigation. 

regarding the low battery RTH I think this should be a separate PR in general but I really like that Idea. some people asked for low battery RTH already but low voltage based RTH makes no sense. But with the now properly working wind estimator, the flight time and flight distance estimation should be able to work out of the box without much tuning (beside idle power). what do you think @shellixyz? We can have windspeed and airspeed now from the AHRS and therefore also proper cruise power and efficiency very precisely. that should make Battery estimation much easier and simpler to set up.

**Comment by shellixyz** [comment]
Yes with accurate wind speed and direction information assuming the idle power, cruise power and cruise speed are set correctly it should be possible to have a reliable RTH based on the estimated time to go home with the specified battery margin which is already implemented. I have tested it a several times in the past without wind and it seemed to be working fine. I haven't had any negative feedback about it either. Although I haven't got any feedback at all so I don't know how many people have used it and it has been a while since I last tested it so I don't know what impact modifitications to the code since I tested it had.

Basically adding the feature would just be a matter of adding a parameter to be able to enable/disable it and when it is enabled and `calculateRemainingEnergyBeforeRTH()` returns 0 trigger RTH.

**Comment by b14ckyy** [comment]
Then I can give you feedback: If properly tuned it works very well. I always set it up a bit more conservative although mostly I had too much battery left after return but as wind speed estimation was never good, this was better than the opposite. 

so an automatic trigger for GPS assisted modes with user override option would be great. Don't yxou think cruise power and cruise speed could be determined automatically now, during level flight in navigation modes? with an average over the last 1 or 2 minutes of flight? 

anyway that's a topic for a different ticket/PR. Just came into my mind.

**Comment by shellixyz** [comment]
Sure it should be possible to auto-learn the cruise parameters fairly easily assuming air speed estimation is reliable

**Comment by RomanLut** [comment]
> I agree on `failsafe_mission_gps_fix_estimation_delay`, this sounds like a good option. But maybe call it in general `gps_fix_estimation_failsafe_delay` and apply it to any navigation mode. So if GPS Fix Estimation has to take over due to a fix loss, the plane can failsafe and RTH in any GPS assisted mode unless the user takes back over with stick input or disables navigation.
> 
RTH is GPS -assisted mode too :) So an option should be related to mission mode configuration only.

**Comment by b14ckyy** [comment]
True, but even during RTH a failsafe can happen. And a failsafe is an indicator that something is wrong. That's why my idea is, to trigger a failsafe condition as soon as the GPS fix is lost. At least as an option. I see no reason why this should only be limited to missions.

**Comment by RomanLut** [comment]
Added failsafe_gps_fix_estimation_delay

Controls whether waypoint mission is allowed to proceed with gps fix estimation. Sets the time delay in seconds between gps fix lost event and RTH activation. Minimum delay is 7 seconds. If set to -1 the mission will continue until the end. With default setting(7), waypoint mission is aborted and switched to RTH with 7 seconds delay. RTH is done with GPS Fix estimation.

**Comment by bakwc** [comment]
Hey! Any chances to get it merged? Very cool feature!

**Review by ABLomas — APPROVED** [review]

**Comment by b14ckyy** [comment]
Not for 6.0. needs thoroughly flight tested first and currently for most of us that's not possible due to weather constrains.

**Comment by ABLomas** [comment]
> Not for 6.0. needs thoroughly flight tested first and currently for most of us that's not possible due to weather constrains.

Sorry, this was mistake with approval from my side (never used vscode plugin before).
As for testing - weather is really bad, but i have code loaded on plane, hope to test it tomorrow.

**Comment by ABLomas** [comment]
Tested today on plane.
Does not work, at all so far ;-)
(in all tested cases triggers EMERGENCY LANDING and does not even try to navigate)
INAV 6.0.0 branch with this patch cleanly applied, settings:

set inav_allow_gps_fix_estimation=ON
set fw_reference_airspeed=1583
set failsafe_gps_fix_estimation_delay=-1

Did simple switch on RC12 to disconnect TX wire from GPS.

RAW DVR: https://youtu.be/thhAlb1ZRNs
Timecodes:
0:27 - disconnected GPS during RTH procedure - goes to RTH.
Then decided to verify compass calibration and waypoints - simple track, square, looks like compass working perfectly, navi works fine.
03:19 disconnected GPS during route to next waypoint. Immediately triggered EMERGENCY LANDING
04:55 - one more disconnect in air, again the same.

Also have blackbox at 3% rate i think if needed.

**Comment by bakwc** [comment]
Do you have compass?

**Comment by ABLomas** [comment]
> Then decided to verify compass calibration and waypoints

"Then decided to verify compass calibration and waypoints"...
So - yes, i have it.

**Comment by b14ckyy** [comment]
Do you have Airspeed sensor? At least virtual Airspeed enabled and calibrated (multiple direction changes in nav/cruise mode).

**Comment by ABLomas** [comment]
> Do you have Airspeed sensor? At least virtual Airspeed enabled and calibrated (multiple direction changes in nav/cruise mode).

No, i do not have airspeed sensor.
In description: "pitot (optional)"
I have set reference airspeed, as described 
(set fw_reference_airspeed=1583)
Is virtual airspeed sensor really required? Why, where is logic in that - we have reliable cruise throttle, we have reference airspeed to calculate distance over time, what added value on virtual airspeed which is not working anyway without GPS?
(and i have I2C airspeed nearby, could attach it if it's really required, but i guess emergency landing is just from additional missed trigger, not due to missing airspeed sensor)

**Comment by ABLomas** [comment]
@b14ckyy if you spotted place in code which absolutely requires airspeed sensor to work - let me know. Tomorrow weather will be acceptable to fly, i can plug I2C airspeed and retry.
I quickly checked code and did not saw such dependency, but maybe i missed it. Not sure what can i do more for now...

**Comment by b14ckyy** [comment]
I just had that remembered that it needs airspeed, but I could be wrong. I always have reference airspeed turned on since 6.0 testing started and on my test platform I have a physical airspeed sensor. So I never tried with airspeed disabled.

**Comment by breadoven** [comment]
One thing I've just realised with this is it doesn't include a Useage #define. I think this needs to be added.

**Comment by ABLomas** [comment]
I tested today with virtual pitot.
Made zero difference, after i killed GPS - it got into emergency landing mode (i was ~700m from home, in RTH mode, so all checks "too close to home", etc should not apply)

**Comment by breadoven** [comment]
> I tested today with virtual pitot. Made zero difference, after i killed GPS - it got into emergency landing mode (i was ~700m from home, in RTH mode, so all checks "too close to home", etc should not apply)

A blackbox log would be the best way of understanding this.

**Comment by ABLomas** [comment]
Maybe not a problem to provide BB log, but hey, it is working for everyone else (who developed this) and only i have problems?
I can't see anything specific in configuration, which could lead to different logic, so if any dev tested this and it's working - maybe share configuration; else if it's not working - it would be much easiers to have "faulty configuration" right here, right now on desk.

(BB log is 45km away, have only DVR for now)

**Comment by ABLomas** [comment]
I also see two big differences:
1) AFAIK nobody else tested it on real craft, only on HILSIM, maybe some conditions differs (inputs from virtual sensors and so on)
2) i see that documentation suggests using IPF to turn off GPS. Maybe it kills GPS in different way than it would be in real world. I used some kind of DIY "kill switch on TX line from GPS" -  physically disconnecting one wire - which excludes possible bugs on IPF implementation side.

Would be nice if someone who really tested this feature could describe their setup.

**Comment by breadoven** [comment]
If you provide a BB log it will show immediately if the Emergency Landing is caused by lack of a valid position or lack of a valid heading. RTH sanity could also cause an Emergency Landing but that seems unlikely.

**Comment by b14ckyy** [comment]
@ABLomas I tested on real aircraft. But with Pitot tube installed. It worked. Wind was very gusty causing quick error buildup but it worked to bring it back by 200m offset from a 2km distance

**Comment by RomanLut** [comment]
I will double check it on theese weekends. Airspeed sensor is not requried. Compass is requried.
Code was written to handle GPS Fix lost reported by GPS sensor, but tx line disconnection is GPS sensor timeout/failure.
I am not sure that code handes this case.

**Comment by ABLomas** [comment]
Oh, so there are different cases i guess.
I can change to different failure type (disconnect GPS antenna) and recheck.
(did today one more flight with "TX line disconnect") and again had no time after to DL blackbox data, will do that tomorrow evening, then we will know how it works).

**Comment by ABLomas** [comment]
[blackbox_log_2023-02-18_155602_df.zip](https://github.com/iNavFlight/inav/files/10774932/blackbox_log_2023-02-18_155602_df.zip)
A bit big log but managed to capture even two GPS failsafes at once:
 - at 1:18 i lost GPS ("naturally" - means TX line not cut, etc - due to bank angle and specific configuration). Aircraft entered emergency landing phase, but GPS signal recovered and mission continued. MAYBE this time mission did not kicked in because i was +/- close to home point, maybe.
 - 14:33 i tested with cut TX line out from GPS, again aircraft entered emergency landing phase.
 
In both cases - no attempt to continue mission/route.

**Comment by breadoven** [comment]
The log shows the position isn't trusted once the GPS is cut (there's a 4-5s timeout delay), hence the emergency landing. RTH/WP mode won't work if the position isn't trusted. Not sure how this change handles that, I assume it should maintain position as trusted even without a valid GPS fix.

File: src/main/flight/failsafe.h
Comment (breadoven) [inline src/main/flight/failsafe.h]: Tabbed indentation ? https://github.com/iNavFlight/inav/blob/master/docs/development/Development.md

Thought the checking system complained about this ?

File: src/main/io/gps.c
Comment (breadoven) [inline src/main/io/gps.c]: Should be conditional on `allow_gps_fix_estimation ` setting or `canEstimateGPSFix `being true surely ? No point calling this if use option not set, isn't a plane, has no mag etc.

File: src/main/navigation/navigation_fixedwing.c
Comment (breadoven) [inline src/main/navigation/navigation_fixedwing.c]: Why has the indentation been shifted ?

File: src/main/navigation/navigation_fixedwing.c
Comment (breadoven) [inline src/main/navigation/navigation_fixedwing.c]: indentation ?

**Review by breadoven — COMMENTED** [review]
This feature should really include a usage Define# in case users prefer not to include it in custom builds.

**Comment by RomanLut** [comment]
Had no time to do all necessary work on this task theese weekends.
Checked RTH and Mission in simulator. GPS was disabled from HITL menu and by Logic Condition - works as intended.
There probably some diiferrence between fix lost reported by GPS, and GPS sensor failure; still to check this.

**Comment by ABLomas** [comment]
> There probably some diiferrence between fix lost reported by GPS, and GPS sensor failure; still to check this.

I attached blackbox log above - even normal "GPS sat lost" triggers emergency landing instead.
So probably there are few more conditions and few differences from HILSIM.

**Comment by RomanLut** [comment]
> I attached blackbox log above - even normal "GPS sat lost" triggers emergency landing instead. So probably there are few more conditions and few differences from HILSIM.
How you build firmware?

**Comment by ABLomas** [comment]
buildflags
WSL + VSCode - selected PR and built.
Patch applied for sure, else new variables like "inav_allow_gps_fix_estimation" and so on won't work at all in CLI.

**Comment by breadoven** [comment]
It's apparent from the DVR recording above (https://github.com/iNavFlight/inav/pull/8347#issuecomment-1429846280) that the OSD Sats field drops to 0 when the GPS is cut followed by the emergency landing several seconds later. The OSD Sats field remains on 0 which doesn't make sense if the GPS has failed. The OSD Sats field should blink and show "X!" in this case. In fact the field should blink anyway with 0 Sats, which it doesn't.

Checking the PR changes again it's obvious the above is caused by a bug in the OSD code which prevents GPS no fix/failure being indicated if `GPS_ESTIMATED_FIX `isn't set. This bug also confirms that `GPS_ESTIMATED_FIX `isn't being set when the GPS serial connection is cut, instead the GPS fails which prevents this option from working because `sensors(SENSOR_GPS)` is no longer set.

**Comment by ABLomas** [comment]
I agree that cutting TX wire from GPS to simulate it's failure is not typical scenario.
But here's short DVR recording for blackbox data above (~ 1:18 in blackbox):
https://user-images.githubusercontent.com/8989151/220082140-af258699-e55a-4202-9f9a-fca12966958a.mp4
Here i lost GPS for "completely natural" causes, did not cut anything and so on, simply lost GPS lock - and instead got emergency landing triggered again (it's visible that aircraft changed it's pitch angle already - not continuing mission).
So there are, i believe, two different issues (and both not visible on HILSIM, but only in reality).

_(let's skip discussion why GPS lost lock so easily, this is expected at such bank angles due to design - it's experimental plane - but can happen in real environment as well)_

**Comment by RomanLut** [comment]
Currently feature does not handle GPS failure.

On 1:18, gp fix estimation works correctly, there is UNUSED flag whish is GPS_ESTIMATED_FIX:
image

On 14:33, there is GPS failure ( no GPS OK state) and no UNUSED (GPS_ESTIMATED_FIX) flag:

image

I have to enhance feature to handle GPS failures too.

**Comment by ABLomas** [comment]
Oh, ok then, i will rework remote switch - to switch RF PA OFF (on GPS), instead of TX line and retry. Sadly, can't re-test tomorrow due to weather - flying in water (snow, rain) is fun and plane can handle it, but if something goes wrong - i can't overtake in LOS, so at tuesday then.
GPS failure itself is rare, would be nice if it can be handled, but lost sat lock is much more common.

Still, why i'm seeing EMERGENCY LANDING on OSD, if this GPS failure was handled correctly, and why this sudden pitch down moment at the same time (looks like plane was going to really land, like expected)? If DR is working - should i still see EMERGENCY LANDING on screen even when it's working properly? Confusing...

**Comment by DangerD1024** [comment]
Чувак, напиши в телегу @dangerd512  =)

**Comment by RomanLut** [comment]
There is still a bug:
if mission is aborted with failsafe_gps_fix_estimation_delay, then landing will be enabled too early, before RTH is fully finished.
Still working on it.

**Comment by RomanLut** [comment]
Fixed. Now everything should work fine.

I still to dedicate time to surround feature code with #ifdef USE_GPS_FIX_ESTIMATION

**Comment by ABLomas** [comment]
Just a comment about real-world test.
Almost entire flight is fine, GPS was jammed by onboard device, flight ongoing with waypoints.
Airplane reached waypoints, then regained GPS lock, got into next waypoint, again lost lock. And then, at some time, it got into EMERGENCY LANDING phase.
I do not have blackbox and i will try to replicate this, but my guess what happens:
- i have `set gps_dyn_model = PEDESTRIAN`
 - it drifted in "fake GPS" mode a lot (due to wind?)
 - GPS filter kicked in and, even with GPS lock enabled and good HDOP - real position was dismissed/filtered

Need to repeat test and bb log this issue.

**Comment by DzikuVx** [comment]
@RomanLut I'm moving this to INAV 7. But I'd like to merge whenever you say it's safe to merge. So, is it safe to merge?

**Comment by ABLomas** [comment]
Comment about reliability:
 - i have this firmware (minus yesterdays merge) installed on 3 planes, one is pretty much complex, with IPF, 8 servos outputs, etc. Total mileage > 600km
 - it works as dead reckoning solution, not precise, but we all understand why. And tries to do mission if GPS is unavailable. At least as "bring home" solution - works pretty well
 - **there is** one remaining issue with triggered emergency landing and i will try to replicate it with blackbox, pretty much edge case (without this patch plane will go emergency landing anyway)

So, while not ideal - it works, it helps a lot, but some more fixes may be required (or additional documentation to avoid those corner cases)

**Comment by RomanLut** [comment]
> @RomanLut I'm moving this to INAV 7. But I'd like to merge whenever you say it's safe to merge. So, is it safe to merge?

@DzikuVx ABLomas reports emergency landing triggerred spontaneously on mission. I think it will be better to delay merge untill I understand what goes wrong.

**Comment by DzikuVx** [comment]
Understood @RomanLut Ping me when you think it's safe enough to proceed

**Comment by Alessbag** [comment]
I tested this in real wing. Sonic model Ar Wing Pro. 5 km distant and enable gps off and activated failsaаe. Wing turn on home point. Еhere was wind so the plane arrived with an error of about 500m.

**Comment by b14ckyy** [comment]
@RomanLut can you fix the conflicts please? I wanna test the latest status in real life so we can get this in for 7.0.

**Comment by RomanLut** [comment]
> @RomanLut can you fix the conflicts please? I wanna test the latest status in real life so we can get this in for 7.0.

Done. It should work. I tested it IRL by 1) disabling GPS by RC switch using programming framework, 2) physically disconnecting GPS after arming (GPS HW failure), 3) with switchable GPS jamming device installed onboard.

**Comment by ABLomas** [comment]
I confirm it working (also tested with jammer and physical switch (relay over PWM)).
I also tried to reproduce emergency landing issue and while i could not reliably trigger it (TBH needs more time to reproduce it and make BB logs) i believe problem is in GPS filter: 

- set GPS filter to PEDESTRIAN (helps agains spoofing on slower planes - yep, not everyone uses AIRBORNE 8G)
- let plane fly autonomously, remove GPS
- after some flight (let's say, 3km without GPS) add GPS back (or, in other words - fly over jammer)
- GPS lock will be regained, but for short time plane will probably enter EMERGENCY LANDING state, as we have GPS lock, but coords disagree a lot (and are filtered on GPS side, while we have many sats available)

Anyway, this is corner case and this patch is very, very useful for many INAV users, i vote for adding it to official release.

**Comment by b14ckyy** [comment]
@RomanLut @ABLomas Sorry that I have to ask but can one of you share a Configurator build that has the disable GPS logic function? I messed up my build system yesterday and still working to get it up again before I reinstall everything. :(

**Comment by RomanLut** [comment]
> @RomanLut @ABLomas Sorry that I have to ask but can one of you share a Configurator build that has the disable GPS logic function? I messed up my build system yesterday and still working to get it up again before I reinstall everything. :(

I do not know how to build executable, I just type "npm start" in the folder.
I think to repair build environment, you have to install nmp, delete node_moduler folder, "npm cache clean --force", "npm install", "npm start".

Anyway, here is a program which disables GPS on Channel 8 High:
logic 0 1 -1 6 1 8 0 0 0
logic 1 1 -1 52 4 0 0 0 0

**Comment by b14ckyy** [comment]
@DzikuVx since this is now also real world tested as well and seems to be usable, I think we can merge it? More people can test it then.

**Comment by breadoven** [comment]
Still needs usage define adding as mentioned above https://github.com/iNavFlight/inav/pull/8347#issuecomment-1461026829.

**Comment by RomanLut** [comment]
> Still needs usage define adding as mentioned above [#8347 (comment)]

Oops I forgot about it, I will do it next week.

**Comment by RomanLut** [comment]
>I also tried to reproduce emergency landing issue and while i could not reliably trigger it (TBH needs more time to reproduce it and make BB logs) i believe problem is in GPS filter:

"GPS filter" (or GPS Glitch detection) has been disabled in inav since 4.0 I think.
I guess Emergency landing is trigerred by "nav_rth_abort_threshold".
"nav_rth_abort_threshold" is disabled while estimation is active, and is kept disabled 5 seconds after GPS Fix is reacquired to avoid false triggering when estimated coordinates are replaced with real coordinates from GPS sensor.
However, if GPS sensor returns highly inacurate coordinates, or is spoofed, RTH will be aborted with emergency landing. Basically, it has nothing to do with GPS fix estimation.
If you experience such problem then disabling nav_rth_abort_threshold should help.

(Also I assume you tested version which includes this commit https://github.com/iNavFlight/inav/pull/8347/commits/4d888e438b542eac3b0ac1206cd87e7894ca8618 This is important fix which fixes all kinds of weird behaviour)

**Comment by RomanLut** [comment]
> @RomanLut I'm moving this to INAV 7. But I'd like to merge whenever you say it's safe to merge. So, is it safe to merge?

Feature is ready to merge now

**Comment by DangerD1024** [comment]
Any workaround with spoofing?

**Comment by DzikuVx** [comment]
@RomanLut thi somehow got below my radar :( Can you resolve conflicts and then we will schedule for 7.1?

**Comment by RomanLut** [comment]
> @RomanLut thi somehow got below my radar :( Can you resolve conflicts and then we will schedule for 7.1?

Resolved

**Comment by RomanLut** [comment]
> @RomanLut damn, we have more conflicts now....

@DzikuVx Resolved

**Comment by bakwc** [comment]
Cool!

**Pull request merged**
